spec/validations/requests_spec.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+def request_function(style, address)
+  case style
+  when 'ZAF'
+    "function request() { client.request('#{address}') }"
+  when 'jQuery'
+    "function request() { $.get('#{address}') }"
+  when 'XMLHttpRequest'
+    "function request() { xhr.open('get', '#{address}') }"
+  when 'fetch'
+    "function request() { fetch('#{address}') }"
+  end
+end
+
+request_function_styles = ['ZAF', 'jQuery', 'XMLHttpRequest', 'fetch']
+
+shared_examples 'an insecure request' do |function_style|
+  let(:markup) { request_function(function_style, 'http://foo.com') }
+
+  it "and raise a warning inside #{function_style} style requests" do
+    # TODO: update with translated string
+    warning = 'Possible insecure HTTP request to http://foo.com in assets/app.js. Consider instead using the HTTPS protocol.'
+    errors = subject.call(package)
+    expect(package.warnings).to include(warning)
+    expect(errors).to be_empty
+  end
+end
+
+blocked_requests = [
+  {
+    type: 'private',
+    example_address: 'https://192.168.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to private network address 192.168.0.1 in assets/app.js'
+  },
+  {
+    type: 'loopback',
+    example_address: 'https://127.0.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to a loopback network address 127.0.0.1 in assets/app.js'
+  },
+  { type: 'link-local',
+    example_address: 'https://169.254.0.1',
+    # TODO: update with translated string
+    error: 'Possible request to a link-local network address 169.254.0.1 in assets/app.js'
+  }
+]
+
+blocked_requests.each do |blocked_request|
+  shared_examples "a blocked #{blocked_request[:type]} request" do |function_style|
+    let(:markup) { request_function(function_style, "#{blocked_request[:example_address]}") }
+
+    it "and error inside #{function_style} style request calls" do
+      errors = subject.call(package)
+      expect(package.warnings).to be_empty
+      expect(errors).to include(blocked_request[:error])
+    end
+  end
+end
+
+describe ZendeskAppsSupport::Validations::Requests do
+  let(:app_js) { double('AppFile', relative_path: 'assets/app.js', read: markup) }
+  let(:subject) { ZendeskAppsSupport::Validations::Requests }
+  let(:package) { double('Package', js_files: [app_js], html_files: [], warnings: []) }
+
+  describe 'using the http protocol' do
```

Comment:
I like how a bunch of this stuff is abstracted into shared examples. I wonder if it can be taken further - something like:
```
request_function_styles.each do |style|
  describe "#{style} type requests" do
    describe 'to some address range' do
      it_behaves_like 'a blocked X request', style
...
```

Reply:
It _could_, but I think we sacrifice a fair bit of readability for the sake of condensing 4 cases into 2 here. On the other hand avoiding `shared_examples` all together pushes this out to a _minimum_ of 16 cases (i.e. if we only test these 4 http function call types and no others).

I actually think it's worth keeping these seperate. What do you think?

Reply:
Hmm actually this might be OK in conjunction with your suggestion below. I'll give this a go.